File: BlueSTSDK/BlueSTSDK/BlueSTSDKManager.m
		/*******************************************************************************
 * COPYRIGHT(c) 2015 STMicroelectronics
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *   1. Redistributions of source code must retain the above copyright notice,
 *      this list of conditions and the following disclaimer.
 *   2. Redistributions in binary form must reproduce the above copyright notice,
 *      this list of conditions and the following disclaimer in the documentation
 *      and/or other materials provided with the distribution.
 *   3. Neither the name of STMicroelectronics nor the names of its contributors
 *      may be used to endorse or promote products derived from this software
 *      without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 ******************************************************************************/

#import <Foundation/Foundation.h>
#import <CoreBluetooth/CoreBluetooth.h>

#import "Util/BlueSTSDKBleNodeDefines.h"
#import "BlueSTSDKManager_prv.h"
#import "BlueSTSDKNode_prv.h"
#import "BlueSTSDKNodeFake.h"

#define RETRAY_START_SCANNING_DELAY (0.5) //half second

@interface BlueSTSDKManager()<CBCentralManagerDelegate>
@end

@implementation BlueSTSDKManager{
    /**
     *  true if the manager is scanning for a new nodes
     */
    bool mIsScanning;
    
    /**
     *  concurrent queue to use for do callback
     */
    dispatch_queue_t mNotificationQueue;
    
    /**
     *  set of BlueSTSDKManagerDelegate
     */
    NSMutableSet *mManagerListener;
    
    /**
     *  set with the discovered nodes, of type BlueSTSDKNode
     */
    NSMutableArray *mDiscoveredNode;
    
    /**
     * contains the map (featureMask_t,Feature class) for each know node id
     */
    NSMutableDictionary *mNodeFeatureMap;
    
    /**
     *  system ble manager
     */
    CBCentralManager * mCBCentralManager;
}


+(instancetype)sharedInstance {
    static BlueSTSDKManager *this = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        this = [[self alloc] init];
    });
    return this;
}

/**
 *  initialize the private variables
 *
 *  @return instance of a class BlueSTSDKManager
 */
-(id)init {
    self = [super init];
    
    mDiscoveredNode = [NSMutableArray array];
    mManagerListener = [NSMutableSet set];
    mCBCentralManager = [[CBCentralManager alloc] initWithDelegate:self queue:nil];
    mNotificationQueue = dispatch_queue_create("BlueSTSDKManager", DISPATCH_QUEUE_CONCURRENT);
    

    NSDictionary *defaultValue =[BlueSTSDKBoardFeatureMap boardFeatureMap];
    mNodeFeatureMap = [NSMutableDictionary dictionaryWithCapacity:defaultValue.count];
    //for each key in defaultValue, add a new entry in the mNodeFeatureMap
    [defaultValue enumerateKeysAndObjectsUsingBlock:^(id key, id object, BOOL *stop) {
        [mNodeFeatureMap setObject:[NSMutableDictionary dictionaryWithDictionary:object]
                             forKey:key];
    }];
    
    return self;
}

-(void) discoveryStart{
    [self discoveryStart:-1];
}//discoveryStart

-(void)discoveryStartObj:(NSNumber*)timeoutMs{
    [self discoveryStart:timeoutMs.intValue];
}

-(void) discoveryStart:(int)timeoutMs{
    NSDictionary *options = [NSDictionary dictionaryWithObjectsAndKeys:
                             [NSNumber numberWithBool:YES],
                             CBCentralManagerScanOptionAllowDuplicatesKey, nil];
    if(mCBCentralManager.state!=CBCentralManagerStatePoweredOn){
        [self performSelector:@selector(discoveryStartObj:)
                   withObject:[NSNumber numberWithInt:timeoutMs]
                   afterDelay:RETRAY_START_SCANNING_DELAY];
        return;
    }
    [mCBCentralManager scanForPeripheralsWithServices:nil options:options];
    [self changeDiscoveryStatus:true];
    NSTimeInterval delay = -1.0f;
    if(timeoutMs>0) {
        delay = (NSTimeInterval)((double)timeoutMs / 1000.0f);
        //if timeoutMs
    }
    if(delay>0)
        [self performSelector:@selector(discoveryStop)
                   withObject:nil
                   afterDelay:delay];
    //else don't stop the discovery
}//discoveryStart

-(void) discoveryStop{
    if(mCBCentralManager.state!=CBCentralManagerStatePoweredOn)
        return;
    [mCBCentralManager stopScan];
    //remove perfom selector if we stop the discovery before the timeout
    [NSObject cancelPreviousPerformRequestsWithTarget:self
                                             selector:@selector(discoveryStop)
                                               object:nil];
    [self changeDiscoveryStatus:false];
}

-(void)resetDiscovery {
    [self resetDiscovery:YES];
}
-(void)resetDiscovery:(BOOL)force {
    if (force) {
        for( BlueSTSDKNode *node in mDiscoveredNode){
            if([node isConnected]){
                [node disconnect];
            }//if
        }//for
    }
    NSMutableArray *removeMe = [NSMutableArray arrayWithCapacity:mDiscoveredNode.count];
    for( BlueSTSDKNode *node in mDiscoveredNode){
        if( ![node isConnected]){
            [removeMe addObject:node];
        }//if
    }//for
    for (BlueSTSDKNode *remove in removeMe){
        [mDiscoveredNode removeObject:remove];
    }//for
    
    
}//resetDiscovery

-(NSArray*) nodes{
    return mDiscoveredNode; //[mDiscoveredNode allObjects];
}

-(void)addDelegate:(id<BlueSTSDKManagerDelegate>)delegate {
    [mManagerListener addObject:delegate];
    
}
-(void)removeDelegate:(id<BlueSTSDKManagerDelegate>)delegate{
    [mManagerListener removeObject:delegate];
}

-(BOOL)isDiscovering{
    return mIsScanning;
}

/**
 *  change the status and call all the delegate for notify a manager status change
 *
 *  @param newStatus new manager status
 */
-(void)changeDiscoveryStatus:(BOOL)newStatus{
    mIsScanning=newStatus;
    //notify the new status to the other listeners
    for (id<BlueSTSDKManagerDelegate> delegate in mManagerListener) {
        dispatch_async(mNotificationQueue, ^{
            [delegate manager:self didChangeDiscovery:newStatus];
        });
    }//for
}//changeDiscoveryStatus

/**
 *  add a new node and call all the delegate for notify the discovery of a new node
 *
 *  @param node new discovered node
 */
-(void)addAndNotifyNewNode:(BlueSTSDKNode *)node{
    [mDiscoveredNode addObject:node];
    for (id<BlueSTSDKManagerDelegate> delegate in mManagerListener) {
        dispatch_async(mNotificationQueue, ^{
            [delegate manager:self didDiscoverNode:node];
        });
    }//for
}//addAndNotifyNewNode

-(BlueSTSDKNode *)nodeWithName:(NSString *)name{
    for (BlueSTSDKNode *node in mDiscoveredNode) {
        if ([name isEqual: node.name]) {
            return node;
        }//if
    }//for
    return nil;
}//nodeWithName

-(BlueSTSDKNode *)nodeWithTag:(NSString *)tag{
    for (BlueSTSDKNode *node in mDiscoveredNode) {
        if ([tag isEqual: node.tag]) {
            return node;
        }//if
    }//for
    return nil;
}//nodeWithTag

/**
 *  add a new node and call all the delegate for notify the discovery of a new node
 *
 *  @param node new discovered node
 */
- (void) addVirtualNode {
    @try {
        BlueSTSDKNode *fakeNode = [[BlueSTSDKNodeFake alloc] init];
        [self addAndNotifyNewNode:fakeNode];
    }
    @catch (NSException *exception) {//not a valid advertise -> avoid to add it
    }
}
/**
 *  for each key of the dictionary it check that it has only one bit to 1
 *
 *  @param features map of <uint32_t,Feature class>
 *
 *  @return true if all the keys have only one bit set to 1
 */
-(bool)checkFeatureMask:(NSDictionary*)features{
    
    for( NSNumber *n in features ){
        uint32_t temp = n.unsignedIntValue;
        
        //http://stackoverflow.com/questions/109023/how-to-count-the-number-of-set-bits-in-a-32-bit-integer
        temp = temp - ((temp >> 1) & 0x55555555);
        temp = (temp & 0x33333333) + ((temp >> 2) & 0x33333333);
        temp = (((temp + (temp >> 4)) & 0x0F0F0F0F) * 0x01010101) >> 24;
        
        if(temp!=1)
            return false;
    }
    return true;
    
}

-(void)addFeatureForNode:(uint8_t)nodeId features:(NSDictionary*)features{

    if(![self checkFeatureMask:features])
        @throw [NSException
                exceptionWithName:@"Invalid feature key data"
                reason:@"the key must have only one bit set to 1"
                userInfo:nil];
    NSMutableDictionary *addToMe = [mNodeFeatureMap objectForKey:
                                  [NSNumber numberWithUnsignedChar:nodeId]];
    if(addToMe==nil){
        [mNodeFeatureMap setObject:[NSMutableDictionary dictionaryWithDictionary:features]
                             forKey:[NSNumber numberWithUnsignedChar:nodeId]];
    }else{
        [addToMe addEntriesFromDictionary:features];
    }
    
}

-(NSDictionary*)getFeaturesForNode:(uint8_t)nodeId{
    return [mNodeFeatureMap objectForKey:
            [NSNumber numberWithUnsignedChar:nodeId]];
}

-(bool)isValidNodeId:(uint8_t)nodeId{
    return [self getFeaturesForNode:nodeId]!=nil;
}

#pragma mark - BlueSTSDKManager(prv)

-(void)connect:(CBPeripheral*)peripheral{
    if(mCBCentralManager.state!=CBCentralManagerStatePoweredOn)
        return;
    [mCBCentralManager connectPeripheral:peripheral options:nil];
}//connect

-(void)disconnect:(CBPeripheral*)peripheral{
    if(mCBCentralManager.state!=CBCentralManagerStatePoweredOn)
        return;
    [mCBCentralManager cancelPeripheralConnection:peripheral];
}//disconnect

#pragma mark - CBCentralManagerDelegate

/**
 * if the peripheral has a valid advertise we build a new node and notify the discovery
 * otherwise we skip it
 * if the node is already discovered we update its rssi value
 */
- (void)centralManager:(CBCentralManager *)central
 didDiscoverPeripheral:(CBPeripheral *)peripheral
     advertisementData:(NSDictionary *)advertisementData
                  RSSI:(NSNumber *)RSSI{
    NSString *tag = peripheral.identifier.UUIDString;
    
    BlueSTSDKNode *node = [self nodeWithTag:tag];
    if(node == nil){
        @try {
            node = [[BlueSTSDKNode alloc] init:peripheral rssi:RSSI
                                   advertise:advertisementData];
            [self addAndNotifyNewNode:node];
        }
        @catch (NSException *exception) {//not a valid advertise -> avoid to add it
        }
        
    }else{
        [node updateRssi:RSSI];
    }//if-else
}

/**
 * when the system stop the discovery process we notify it to the user
 */
- (void)centralManagerDidUpdateState:(CBCentralManager *)central{
    CBCentralManagerState state = [central state];
    if(state!=CBCentralManagerStatePoweredOn){
        [self changeDiscoveryStatus:false];
    }else{
        [self changeDiscoveryStatus:true];
    }//if-else
}//centralManagerDidUpdateState

/**
 * when the peripheral is connected we call the method
 * {@link BlueSTSDKNode#completeConnection} of the node class
 */
- (void)centralManager:(CBCentralManager *)central
  didConnectPeripheral:(CBPeripheral *)peripheral{
    NSString *tag = peripheral.identifier.UUIDString;
    BlueSTSDKNode *node = [self nodeWithTag:tag];
    if(node == nil) //we did not handle this peripheral
        return;
    [node completeConnection];
}//didConnectPeripheral

/**
 * when a connection fail we call the method {@link BlueSTSDKNode#connectionError:}
 * of the node class
 */
-(void)notifyConnectionError:(CBPeripheral*)peripheral error:(NSError*)error{
    NSString *tag = peripheral.identifier.UUIDString;
    BlueSTSDKNode *node = [self nodeWithTag:tag];
    if(node == nil) //we did not handle this peripheral
        return;
    [node connectionError:error];
}//notifyConnectionError

/**
 * when a connection fail we call the method {@link BlueSTSDKNode#connectionError:}
 * of the node class
 */
- (void)centralManager:(CBCentralManager *)central
didFailToConnectPeripheral:(CBPeripheral *)peripheral
                 error:(NSError *)error{
    [self notifyConnectionError:peripheral error:error];
}//didFailToConnectPeripheral

/**
 * when the peripheral is disconnected we call the method 
 * {@link BlueSTSDKNode#completeDisconnection:} of the node class
 */
- (void)centralManager:(CBCentralManager *)central
didDisconnectPeripheral:(CBPeripheral *)peripheral
                 error:(NSError *)error{
    NSString *tag = peripheral.identifier.UUIDString;
    BlueSTSDKNode *node = [self nodeWithTag:tag];
    if(node == nil) //we did not handle this peripheral
        return;
    [node completeDisconnection:error];
}//didDisconnectPeripheral

@end

File: BlueSTSDK/BlueSTSDK/BlueSTSDKManager.swift
import Foundation

@objc public class BlueSTSDKManager : NSObject{
    private static let RETRAY_START_SCANNING_DELAY = TimeInterval(0.5)
    public static let DEFAULT_ADVERTISE_FILTER:[BlueSTSDKAdvertiseFilter] = [BlueSTSDKAdvertiseParser()];
    
    @objc public static let sharedInstance = BlueSTSDKManager()
    /**
     *  Get all the discovered nodes
     *
     *  @return array of {@link BlueSTSDKNode} with all the discovered nodes
     */
    @objc private(set) public var nodes:[BlueSTSDKNode] = []
    
    /**
     *  Tell if the manager is in a discovery state
     *
     *  @return true if the manager is seaching for new nodes
     */
    @objc private(set) public var isDiscovering:Bool = false {
        didSet{
            mListeners.forEach{ listener in
                mNotificationQueue.async {
                    listener.manager(self, didChangeDiscovery: self.isDiscovering)
                }
            }
        }
    }
    
    private var mListeners:[BlueSTSDKManagerDelegate] = []
    private let mNotificationQueue = DispatchQueue(label: "BlueSTSDKMenager",qos: .background, attributes: .concurrent)
    private var mNodeFeatures:[UInt8:[NSNumber:BlueSTSDKFeature.Type]] = [:]
    private var mCBCentralManager:CBCentralManager!
    private var mAdvertiseFilters:[BlueSTSDKAdvertiseFilter]=BlueSTSDKManager.DEFAULT_ADVERTISE_FILTER
    
    private override init() {
        super.init()
        mCBCentralManager = CBCentralManager(delegate: self, queue: nil)
        let  defaultFeatureMap = BlueSTSDKBoardFeatureMap.boardFeatureMap
        defaultFeatureMap.forEach{ boardId, featureMap in
            mNodeFeatures[boardId.uint8Value] = featureMap
        }
    }
    
    @objc public func discoveryStart(){
        discoveryStart(-1)
    }
    
    @objc public func discoveryStart(_ timeoutMs:Int ){
        discoveryStart(timeoutMs, advertiseFilters: BlueSTSDKManager.DEFAULT_ADVERTISE_FILTER)
    }
    /**
     *  Start a discovery process that will scan for a new node. The discovery will
     * stop after {@code timeoutMs} milliseconds
     *
     *  @param timeoutMs milliseconds to wait before stop the scanning
     */
    @objc public func discoveryStart(_ timeoutMs:Int, advertiseFilters: [BlueSTSDKAdvertiseFilter] ){
        guard mCBCentralManager.state == .poweredOn  && isDiscovering == false else{
            mNotificationQueue.asyncAfter(deadline: .now()+BlueSTSDKManager.RETRAY_START_SCANNING_DELAY){
                self.discoveryStart(timeoutMs,advertiseFilters: advertiseFilters)
            }
            return
        }
        let scanOptions:[String : Any] = [CBCentralManagerScanOptionAllowDuplicatesKey:true]
        mAdvertiseFilters = advertiseFilters
        mCBCentralManager.scanForPeripherals(withServices: nil, options: scanOptions)
        isDiscovering = true
        let delay = TimeInterval(timeoutMs)/1000.0
        if(delay>0){
            mNotificationQueue.asyncAfter(deadline: .now() + delay){
                self.discoveryStop()
            }
        }
    }
    
    /**
     *  Stop the discovery process
     */
    @objc public func discoveryStop(){
        guard mCBCentralManager.state == .poweredOn else {
            return
        }
        guard isDiscovering == true else{
            return
        }
        mCBCentralManager.stopScan()
        isDiscovering = false
    }
 
    //TODO: swift: use defailt parameters
    @objc public func resetDiscovery(){
        resetDiscovery(true)
    }
    /**
     *  Remove all the discovered nodes
     *  @param force if true remove also the connected nodes
     */
    @objc public func resetDiscovery(_ disconnectAll:Bool){
        if(disconnectAll){
            nodes.forEach{ node in
                if(node.isConnected()){
                    node.disconnect()
                }
            }
        }
        //remove all disconnected node
        nodes.removeAll{ node in
            !node.isConnected()
        }
    }
    
    fileprivate func addAndNotify(node: BlueSTSDKNode){
        guard !nodes.contains(node) else{
            return
        }
        nodes.append(node)
        mListeners.forEach{ listener in
            mNotificationQueue.async {
                listener.manager(self, didDiscoverNode: node)
            }
        }
    }
    
    /**
     *  add a new node and call all the delegate for notify the discovery of a new node
     *
     *  @param node new discovered node
     */
    @objc public func addVirtualNode(){
        if let node = BlueSTSDKNodeFake(){
            addAndNotify(node: node)
        }
    }
    
    /**
     *  Search in the discovered node the one that has a particular name,
     *  @note the node name is not unique so we will return the first node that match the name
     *
     *  @param name node name to search
     *
     *  @return a node with the name, or nil if a node with that name doesn't exist
     */
    @objc public func nodeWith( name:String)->BlueSTSDKNode?{
        return nodes.first{ $0.name == name}
    }
    
    /**
     *  Search in the discovered node the one that has a particular tag
     *
     *  @param tag tag to search
     *
     *  @return a node with that tag or nil if the node doesn't exist
     */
    public func nodeWith(tag:String) -> BlueSTSDKNode?{
        return nodes.first{ $0.tag == tag}
    }
 
    private func isValidFeatureMap( featureMap: [UInt32:BlueSTSDKFeature.Type])->Bool{
        return featureMap.keys.allSatisfy{ $0.isValidFeatureMask }
    }
    /**
     *  Add to a node a list of possible feature exported by the node
     *
     *  @param nodeId  new board id or a board id that we want extend
     *  @param features map of new features add to the board, it is a dictionary of
     * <{@link featureMask_t},BlueSTSDKFeature>
     * @throw an exception if the featureMask as more than one bit set to 1
     */
    @objc public func addFeatureForNode(nodeId:UInt8, features:[UInt32:BlueSTSDKFeature.Type]) throws{
        guard isValidFeatureMap(featureMap: features) else{
            NSException(name: NSExceptionName("Invalid feature key data"), reason: "the key must have only one bit set to 1", userInfo: nil).raise()
            return
        }
        
        var currentMap = mNodeFeatures[nodeId] ?? BlueSTSDKBoardFeatureMap.defaultMaskToFeatureMap
        
        features.forEach{ (arg) in
            let (key, feature) = arg
            currentMap[NSNumber(value: key)] = feature
        }
        mNodeFeatures[nodeId] = currentMap
    }
    
    @objc public func getFeaturesForNode(_ nodeId:UInt8)-> [NSNumber : BlueSTSDKFeature.Type]{
        if let featuresMap = mNodeFeatures[nodeId] {
            return featuresMap
        }else {
            return BlueSTSDKBoardFeatureMap.defaultMaskToFeatureMap
        }
        
    }
    
    @objc public func removeDelegate(_ delegate:BlueSTSDKManagerDelegate){
        
        if let index = mListeners.firstIndex(where: {$0 === delegate}){
            mListeners.remove(at: index)
        }
    }
    
    @available(*, deprecated, message: "Use removeDelegate")
    @objc public func remove(_ delegate:BlueSTSDKManagerDelegate){
        removeDelegate(delegate)
    }
    
    @objc public func addDelegate(_ delegate:BlueSTSDKManagerDelegate){
        guard mListeners.firstIndex(where: {$0 === delegate}) == nil else{
            return
        }
        mListeners.append(delegate)
    }
    
    @available(*, deprecated, message: "Use addDelegate")
    @objc public func add(_ delegate:BlueSTSDKManagerDelegate){
        addDelegate(delegate)
    }
    
    //TODO: swift: remove the public when node is rewrited in swift
    @objc public func connect(_ peripheral:CBPeripheral){
        guard mCBCentralManager.state == .poweredOn else{
            return
        }
        mCBCentralManager.connect(peripheral, options: nil)
    }
    
    //TODO: swift: remove the public when node is rewrited in swift
    @objc public func disconnect(_ peripheral:CBPeripheral){
        guard mCBCentralManager.state == .poweredOn else{
            return
        }
        mCBCentralManager.cancelPeripheralConnection(peripheral)
    }
}

@objc public protocol BlueSTSDKManagerDelegate : class {
    /**
     *  Function called when a new node is discovered
     *
     *  @param manager manager that discovered the node (the manger is a singleton,
     *    so this parameter is only for have a consistent method sign with the others delegate)
     *  @param node new node discovered
     */
    func manager(_ manager:BlueSTSDKManager, didDiscoverNode:BlueSTSDKNode)
    
    /**
     *  Function called when the status of the manager change
     *
     *  @param manager manager that discovered the node (the manger is a singleton,
     *    so this parameter is only for have a consistent method sign with the others delegate)
     *  @param enable true if the manger start a scan process, false if it end the scanning
     */
    func manager(_ manager:BlueSTSDKManager, didChangeDiscovery:Bool)
}

extension BlueSTSDKManager : CBCentralManagerDelegate {
    
    
    public func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String : Any], rssi RSSI: NSNumber) {
        let tag = peripheral.identifier.uuidString
        
        
        if let node = nodeWith(tag: tag){
            node.updateRssi(RSSI)
        }else{
            //get the fist value != nil returned by a filter
            let firstMatch = mAdvertiseFilters.lazy.compactMap{ $0.filter(advertisementData)}.first
            if let info = firstMatch{
                let newNode = BlueSTSDKNode(peripheral, rssi: RSSI, advertiseInfo:info)
                addAndNotify(node: newNode)
            }
        }
    }
    
    public func centralManagerDidUpdateState(_ central: CBCentralManager){
        if(central.state != .poweredOn){
            isDiscovering = false
        }
    }
    
    public func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        let tag = peripheral.identifier.uuidString
        if let node = nodeWith(tag: tag){
            node.completeConnection()
        }
    }
    
    public func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
        let tag = peripheral.identifier.uuidString
        if let node = nodeWith(tag: tag){
            node.connectionError(error)
        }
    }
    
    public func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
        let tag = peripheral.identifier.uuidString
        if let node = nodeWith(tag: tag){
            node.completeDisconnection(error)
        }
    }
}

fileprivate extension UInt32 {
    
    var isValidFeatureMask:Bool{
        get{
            return self.nonzeroBitCount == 1
        }
    }
    
}
